### ISSUE TITLE
Hostname module not permanent when cloud-init is involved

### ISSUE THREAD

**Issue by Fale** [ansible/ansible#19711]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME
hostname_module

##### ANSIBLE VERSION

```
2.2.0
```

##### OS / ENVIRONMENT
CentOS 7

##### SUMMARY
The hostname is not permanent and there is no option to make it permanent

##### STEPS TO REPRODUCE
1.
```
- name: Set hostname
  hostname:
    name: my_host.example.com 
```
2. Reboot

##### EXPECTED RESULTS
Machine keeps hostname

##### ACTUAL RESULTS
Machine loose hostaname

**Comment by ansibot** [comment]
cc @alikins, @saito-hideki
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by georgepsarakis** [comment]
Seems that something changed in Centos7: 

http://unix.stackexchange.com/questions/169067/centos-7-hostname-will-not-change 

and the hostname is no longer read from `/etc/sysconfig/network`.

**Comment by Fale** [comment]
Yes, CentOS 7 uses systemd and hostnamed/hostnamectl instead of the old ways to manage hostnames. In the module description there is written that systemd (and therefore hostnamed) are supported (so I would expect it to work)

**Comment by georgepsarakis** [comment]
Yes, it does support it. If you notice in the module code, [here](https://github.com/ansible/ansible-modules-core/blob/devel/system/hostname.py#L115), it tries to auto-detect that the system is managed by systemd. Otherwise it falls back to the `RedhatStrategy`, assigned to the CentOS operating system.

Perhaps this is a problem with the utility function [is_systemd_managed](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/facts.py#L522) ?

**Comment by bcoca** [comment]
is this machine a container?

**Comment by Fale** [comment]
Nope, standard CentOS on Amazon EC2

**Comment by Fale** [comment]
To be even more precise, I use the AMI ami-e68f82fb (eu-central-1) from [here](https://aws.amazon.com/marketplace/pp/B00O7WM7QW)

**Comment by georgepsarakis** [comment]
@Fale can you please execute the following commands on the VM:

- `which systemctl`
- `stat /run/systemd/system/`
- `stat /dev/.run/systemd/`
- `stat /dev/.systemd/`

**Comment by Fale** [comment]
@georgepsarakis no problem.

```
[f.locati@ip-10-1-0-84 ~]$ which systemctl
/usr/bin/systemctl
[f.locati@ip-10-1-0-84 ~]$ stat /run/systemd/system/
  File: ‘/run/systemd/system/’
  Size: 200       	Blocks: 0          IO Block: 4096   directory
Device: 12h/18d	Inode: 7370        Links: 6
Access: (0755/drwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Context: system_u:object_r:systemd_unit_file_t:s0
Access: 2016-12-28 12:04:30.238000000 +0000
Modify: 2017-01-08 10:52:54.069587237 +0000
Change: 2017-01-08 10:52:54.069587237 +0000
 Birth: -
[f.locati@ip-10-1-0-84 ~]$ stat /dev/.run/systemd/
stat: cannot stat ‘/dev/.run/systemd/’: No such file or directory
[f.locati@ip-10-1-0-84 ~]$ stat /dev/.systemd/
stat: cannot stat ‘/dev/.systemd/’: No such file or directory
[f.locati@ip-10-1-0-84 ~]$ 
```

**Comment by georgepsarakis** [comment]
This should be the equivalent of the applied module logic in Bash commands:

```bash
NEW_NAME='myserver'
CURRENT_NAME=`hostname`

if [ "$CURRENT_NAME" != "$NEW_NAME" ]; then
  hostnamectl --transient set-hostname "$NEW_NAME"
fi

CURRENT_PERMANENT_NAME=`hostnamectl --static status`
if [ "$CURRENT_PERMANENT_NAME" != "$NEW_NAME" ]; then
  hostnamectl --pretty set-hostname "$NEW_NAME"
fi
```

Can you test the above?

**Comment by Fale** [comment]
It seems to not work:

```
[root@ip-10-1-0-84 f.locati]# cat hntest.sh 
#!/bin/bash
NEW_NAME='bastion01'
CURRENT_NAME=`hostname`

if [ "$CURRENT_NAME" != "$NEW_NAME" ]; then
  hostnamectl --transient set-hostname "$NEW_NAME"
fi

CURRENT_PERMANENT_NAME=`hostnamectl --static status`
if [ "$CURRENT_PERMANENT_NAME" != "$NEW_NAME" ]; then
  hostnamectl --pretty set-hostname "$NEW_NAME"
fi
[root@ip-10-1-0-84 f.locati]# hostname
ip-10-1-0-84.eu-central-1.compute.internal
[root@ip-10-1-0-84 f.locati]# ./hntest.sh 
[root@ip-10-1-0-84 f.locati]# hostname
ip-10-1-0-84.eu-central-1.compute.internal
```

This is different from what Ansible obtains:
On the bastion machine:
```
[root@ip-10-1-0-84 f.locati]# hostname
ip-10-1-0-84.eu-central-1.compute.internal
```

On the ansible machine:
```
[ansible@ip-10-1-2-175 ansible]$ ansible bastion -i hosts --become -m hostname -a "name=bastion01"
bastion01.otelia.biz | SUCCESS => {
    "ansible_facts": {
        "ansible_domain": "eu-central-1.compute.internal", 
        "ansible_fqdn": "ip-10-1-0-84.eu-central-1.compute.internal", 
        "ansible_hostname": "bastion01", 
        "ansible_nodename": "bastion01"
    }, 
    "changed": true, 
    "name": "bastion01"
}
```

Back on the bastion host:
```
[root@ip-10-1-0-84 f.locati]# hostname
bastion01
[root@ip-10-1-0-84 f.locati]# hostnamectl
   Static hostname: bastion01
         Icon name: computer-vm
           Chassis: vm
        Machine ID: f9afeb75a5a382dce8269887a67fbf58
           Boot ID: ba0eff33fbd54197a4c43c1f4751e75a
    Virtualization: xen
  Operating System: CentOS Linux 7 (Core)
       CPE OS Name: cpe:/o:centos:centos:7
            Kernel: Linux 3.10.0-514.2.2.el7.x86_64
      Architecture: x86-64
[root@ip-10-1-0-84 f.locati]# reboot
[f.locati@ip-10-1-0-84 ~]$ hostname
ip-10-1-0-84.eu-central-1.compute.internal
[f.locati@ip-10-1-0-84 ~]$ hostnamectl
   Static hostname: ip-10-1-0-84.eu-central-1.compute.internal
         Icon name: computer-vm
           Chassis: vm
        Machine ID: f9afeb75a5a382dce8269887a67fbf58
           Boot ID: 5a2910395aba481dac9a717be6fbccd6
    Virtualization: xen
  Operating System: CentOS Linux 7 (Core)
       CPE OS Name: cpe:/o:centos:centos:7
            Kernel: Linux 3.10.0-514.2.2.el7.x86_64
      Architecture: x86-64
```

**Comment by georgepsarakis** [comment]
Thanks. Can you run it once more with 
```bash
/bin/bash -x hntest.sh
```

to see if it actually runs the final command which should set the permanent hostname?

**Comment by Fale** [comment]
PS: the same occurs also with fqdn:

== Bash
```
[root@ip-10-1-0-84 f.locati]# cat hntest.sh 
#!/bin/bash
NEW_NAME='bastion01.example.com'
CURRENT_NAME=`hostname`

if [ "$CURRENT_NAME" != "$NEW_NAME" ]; then
  hostnamectl --transient set-hostname "$NEW_NAME"
fi

CURRENT_PERMANENT_NAME=`hostnamectl --static status`
if [ "$CURRENT_PERMANENT_NAME" != "$NEW_NAME" ]; then
  hostnamectl --pretty set-hostname "$NEW_NAME"
fi
[root@ip-10-1-0-84 f.locati]# hostname
ip-10-1-0-84.eu-central-1.compute.internal
[root@ip-10-1-0-84 f.locati]# hostnamectl
   Static hostname: ip-10-1-0-84.eu-central-1.compute.internal
         Icon name: computer-vm
           Chassis: vm
        Machine ID: f9afeb75a5a382dce8269887a67fbf58
           Boot ID: 7f99a1f194f143afa95644f04159d168
    Virtualization: xen
  Operating System: CentOS Linux 7 (Core)
       CPE OS Name: cpe:/o:centos:centos:7
            Kernel: Linux 3.10.0-514.2.2.el7.x86_64
      Architecture: x86-64
[root@ip-10-1-0-84 f.locati]# ./hntest.sh 
[root@ip-10-1-0-84 f.locati]# hostname
ip-10-1-0-84.eu-central-1.compute.internal
[root@ip-10-1-0-84 f.locati]# hostnamectl 
   Static hostname: ip-10-1-0-84.eu-central-1.compute.internal
   Pretty hostname: bastion01.example.com
Transient hostname: bastion01.example.com
         Icon name: computer-vm
           Chassis: vm
        Machine ID: f9afeb75a5a382dce8269887a67fbf58
           Boot ID: 7f99a1f194f143afa95644f04159d168
    Virtualization: xen
  Operating System: CentOS Linux 7 (Core)
       CPE OS Name: cpe:/o:centos:centos:7
            Kernel: Linux 3.10.0-514.2.2.el7.x86_64
      Architecture: x86-64
```

== Ansible
```
[f.locati@ip-10-1-0-84 ~]$ hostname
ip-10-1-0-84.eu-central-1.compute.internal
[f.locati@ip-10-1-0-84 ~]$ hostnamectl
   Static hostname: ip-10-1-0-84.eu-central-1.compute.internal
         Icon name: computer-vm
           Chassis: vm
        Machine ID: f9afeb75a5a382dce8269887a67fbf58
           Boot ID: 5a2910395aba481dac9a717be6fbccd6
    Virtualization: xen
  Operating System: CentOS Linux 7 (Core)
       CPE OS Name: cpe:/o:centos:centos:7
            Kernel: Linux 3.10.0-514.2.2.el7.x86_64
      Architecture: x86-64
```
```
[ansible@ip-10-1-2-175 ansible]$ ansible bastion -i hosts --become -m hostname -a "name=bastion01.example.com"
bastion01.otelia.biz | SUCCESS => {
    "ansible_facts": {
        "ansible_domain": "example.com", 
        "ansible_fqdn": "bastion01.example.com", 
        "ansible_hostname": "bastion01", 
        "ansible_nodename": "bastion01.example.com"
    }, 
    "changed": true, 
    "name": "bastion01.example.com"
}
```
```
[f.locati@ip-10-1-0-84 ~]$ hostname
bastion01.example.com
[f.locati@ip-10-1-0-84 ~]$ hostnamectl
   Static hostname: bastion01.example.com
         Icon name: computer-vm
           Chassis: vm
        Machine ID: f9afeb75a5a382dce8269887a67fbf58
           Boot ID: 5a2910395aba481dac9a717be6fbccd6
    Virtualization: xen
  Operating System: CentOS Linux 7 (Core)
       CPE OS Name: cpe:/o:centos:centos:7
            Kernel: Linux 3.10.0-514.2.2.el7.x86_64
      Architecture: x86-64
[f.locati@ip-10-1-0-84 ~]$ sudo reboot
[f.locati@ip-10-1-0-84 ~]$ hostname
ip-10-1-0-84.eu-central-1.compute.internal
[f.locati@ip-10-1-0-84 ~]$ hostnamectl
   Static hostname: ip-10-1-0-84.eu-central-1.compute.internal
         Icon name: computer-vm
           Chassis: vm
        Machine ID: f9afeb75a5a382dce8269887a67fbf58
           Boot ID: 7f99a1f194f143afa95644f04159d168
    Virtualization: xen
  Operating System: CentOS Linux 7 (Core)
       CPE OS Name: cpe:/o:centos:centos:7
            Kernel: Linux 3.10.0-514.2.2.el7.x86_64
      Architecture: x86-64
```

**Comment by Fale** [comment]
```
[root@ip-10-1-0-84 f.locati]# hostname
ip-10-1-0-84.eu-central-1.compute.internal
[root@ip-10-1-0-84 f.locati]# hostnamectl 
   Static hostname: ip-10-1-0-84.eu-central-1.compute.internal
   Pretty hostname: bastion01.example.com
         Icon name: computer-vm
           Chassis: vm
        Machine ID: f9afeb75a5a382dce8269887a67fbf58
           Boot ID: 7f99a1f194f143afa95644f04159d168
    Virtualization: xen
  Operating System: CentOS Linux 7 (Core)
       CPE OS Name: cpe:/o:centos:centos:7
            Kernel: Linux 3.10.0-514.2.2.el7.x86_64
      Architecture: x86-64
[root@ip-10-1-0-84 f.locati]# /bin/bash -x hntest.sh
+ NEW_NAME=bastion01.example.com
++ hostname
+ CURRENT_NAME=ip-10-1-0-84.eu-central-1.compute.internal
+ '[' ip-10-1-0-84.eu-central-1.compute.internal '!=' bastion01.example.com ']'
+ hostnamectl --transient set-hostname bastion01.example.com
++ hostnamectl --static status
+ CURRENT_PERMANENT_NAME=ip-10-1-0-84.eu-central-1.compute.internal
+ '[' ip-10-1-0-84.eu-central-1.compute.internal '!=' bastion01.example.com ']'
+ hostnamectl --pretty set-hostname bastion01.example.com
[root@ip-10-1-0-84 f.locati]# hostname
ip-10-1-0-84.eu-central-1.compute.internal
[root@ip-10-1-0-84 f.locati]# hostnamectl 
   Static hostname: ip-10-1-0-84.eu-central-1.compute.internal
   Pretty hostname: bastion01.example.com
Transient hostname: bastion01.example.com
         Icon name: computer-vm
           Chassis: vm
        Machine ID: f9afeb75a5a382dce8269887a67fbf58
           Boot ID: 7f99a1f194f143afa95644f04159d168
    Virtualization: xen
  Operating System: CentOS Linux 7 (Core)
       CPE OS Name: cpe:/o:centos:centos:7
            Kernel: Linux 3.10.0-514.2.2.el7.x86_64
      Architecture: x86-64
[root@ip-10-1-0-84 f.locati]# 
```

**Comment by georgepsarakis** [comment]
I believe this line:

```
hostnamectl --pretty set-hostname "$NEW_NAME"
```

should become:

```
hostnamectl --static set-hostname "$NEW_NAME"
```

since the latter actually writes to `/etc/hostname` and persists the hostname change.


However, it may be EC2/cloud-init specific. I found this [answer](https://ask.fedoraproject.org/en/question/37413/correctly-setting-the-hostname-fedora-20-on-amazon-ec2/?answer=37746#post-id-37746) that applies to Centos7 if you check the comments.

**Comment by Fale** [comment]
```
[f.locati@ip-10-1-0-84 ~]$ hostname
ip-10-1-0-84.eu-central-1.compute.internal
[f.locati@ip-10-1-0-84 ~]$ hostnamectl 
   Static hostname: ip-10-1-0-84.eu-central-1.compute.internal
   Pretty hostname: bastion01.example.com
         Icon name: computer-vm
           Chassis: vm
        Machine ID: f9afeb75a5a382dce8269887a67fbf58
           Boot ID: 7f99a1f194f143afa95644f04159d168
    Virtualization: xen
  Operating System: CentOS Linux 7 (Core)
       CPE OS Name: cpe:/o:centos:centos:7
            Kernel: Linux 3.10.0-514.2.2.el7.x86_64
      Architecture: x86-64
[f.locati@ip-10-1-0-84 ~]$ sudo hostnamectl --static set-hostname "bastion01.example.com"
[f.locati@ip-10-1-0-84 ~]$ hostname
bastion01.example.com
[f.locati@ip-10-1-0-84 ~]$ hostnamectl 
   Static hostname: bastion01.example.com
Transient hostname: ip-10-1-0-84.eu-central-1.compute.internal
         Icon name: computer-vm
           Chassis: vm
        Machine ID: f9afeb75a5a382dce8269887a67fbf58
           Boot ID: 7f99a1f194f143afa95644f04159d168
    Virtualization: xen
  Operating System: CentOS Linux 7 (Core)
       CPE OS Name: cpe:/o:centos:centos:7
            Kernel: Linux 3.10.0-514.2.2.el7.x86_64
      Architecture: x86-64
[f.locati@ip-10-1-0-84 ~]$ sudo reboot
[f.locati@ip-10-1-0-84 ~]$ hostname
ip-10-1-0-84.eu-central-1.compute.internal
[f.locati@ip-10-1-0-84 ~]$ hostnamectl 
   Static hostname: ip-10-1-0-84.eu-central-1.compute.internal
         Icon name: computer-vm
           Chassis: vm
        Machine ID: f9afeb75a5a382dce8269887a67fbf58
           Boot ID: 27412e3893c84f07bb0eab435a4417cf
    Virtualization: xen
  Operating System: CentOS Linux 7 (Core)
       CPE OS Name: cpe:/o:centos:centos:7
            Kernel: Linux 3.10.0-514.2.2.el7.x86_64
      Architecture: x86-64
```

It seems that is still not enough :(.
It seems like `/etc/hostname` gets rewritten at every boot

**Comment by sivel** [comment]
According to https://aws.amazon.com/premiumsupport/knowledge-center/linux-static-hostname-rhel7-centos7/ the step that may likely be missing is:

```
4. Update the /etc/cloud/cloud.cfg file on your RHEL 7 or Centos 7 Linux instance.
     sudo vim /etc/cloud/cloud.cfg
Append the following string at the bottom of the file to ensure that the hostname is preserved between restarts/reboots.
     preserve_hostname: true
Save and exit the vim editor.
```

I am not sure that the hostname module should be responsible for updating `/etc/cloud/cloud.cfg`, you should probably add another task that is responsible making that change.

**Comment by georgepsarakis** [comment]
@sivel yes it seems EC2-specific. It is also mentioned in the StackExchange answer that I referenced in my previous comment.

**Comment by Fale** [comment]
Thanks guys.
I wonder if the Ansible module should detect if it's running on ec2 and do the heavy lifting in that case or not.
I think if we decide that this module is not the right location to fix this, we should at least document this EC2 weirdness in the Ansible hostname module page

**Comment by angystardust** [comment]
Any news about this issue? We also had the same problem today on a el7 aws instance...

**Comment by Fale** [comment]
@angystardust I guess it will be fixed with an amendment to the doc

**Comment by jtognazzi** [comment]
I had a similar problem with a cloudstack template using [cloud-init](https://cloudinit.readthedocs.io/en/latest/). I would say it's not amazon specific, but can happen if cloud-init is used and not configured properly.

**Comment by Fale** [comment]
The default cloud-init configuration of RHEL and CentOS gives this problem (and probably other distros as well)

**Comment by billryan** [comment]
I think it's easy to set static hostname with `lineinfile` module.

```
- name: set hostname
  hostname:
    name={{ inventory_hostname.split('.')[0] }}

- name: disable updating hostname in xcloud
  lineinfile:
    dest: /etc/cloud/cloud.cfg
    regexp: 'update_hostname$'
    line: '# - update_hostname'
    backup: yes
  tags: xcloud
```

**Comment by Fale** [comment]
Yes it is easy, but I think we should decide which is the correct behavior of the hostname module and change the code and/or the documentation to reflect the decision

**Comment by b-long** [comment]
So, would the next step be identifying the current Ansible behavior in `cloud-init` environments?  Then, the Ansible team could know for sure if this is a feature or a bug.  If it's expected behavior, maybe a warning could be added?

One other comment, I was wondering how Ansible currently tests against `cloud-init` environments.  Am I missing anything beyond this: https://github.com/ansible/ansible/search?q=cloud-init&type=Code&utf8=%E2%9C%93 ?

**Comment by ScubaDrew** [comment]
Ran into this today. Would be nice if it detected the default AWS cloud-init environments since they are so common.

**Comment by abadger** [comment]
needs_info

**Comment by abadger** [comment]
Last test was done without cloud-init... Going to test again.

**Comment by Dmitry1987** [comment]
omg is this for real? 🥇   guys, we're in a year 2018, and the core Ansible "hostname" module will have its node name overridden by cloud-init the next time instance boots? LOL, gotta test this now and see if a hostname remains the same :D

**Comment by Dmitry1987** [comment]
it was ok, my hostnames set by ansible on EC2 machines with 'hostname' module did not revert back to those that cloud-init sets on fresh machine, so that seems to work fine 👍

**Comment by cornfeedhobo** [comment]
@Dmitry1987  cloud-init runs on every boot. could you reboot the instance and verify?

**Comment by Dmitry1987** [comment]
@cornfeedhobo yeah, that's what I did, rebooted 2 of my consul master instances to verify I'll not get a new hostname in the service after reboot :) it was fine, hostname remained as set by Ansible, and not reverted to usual "ip-10-12-23-23" format

**Comment by Dmitry1987** [comment]
@ScubaDrew  what is your Ansible version? I use "2.4.3.0", you might try to upgrade or something. For me it worked yesterday, I use ec2 in Oregon.

**Comment by cornfeedhobo** [comment]
@Dmitry1987 Oh, maybe that is the caveat, it sounds like you were expecting the `usual "ip-10-12-23-23" format`. I think this would require you to be using the [cloud-init hostname](http://cloudinit.readthedocs.io/en/latest/topics/modules.html#set-hostname) option. That is the only time I hit this issue.

```
#cloud-config
fqdn: "awesome.example.com"
hostname: "awesome"
```

Because of the nature of cloud-init and running on restart, I am not sure this can be considered an "ansible problem" and might need to be brought up to [cloud-init](https://git.launchpad.net/cloud-init) (I have no idea how to do that).

**Comment by trinitronx** [comment]
Since this has caused myself and others so much frustration & sadness...

Here's the workaround hack that I've been using:

- `playbooks/example/templates/user-data.sh.j2`:

```
#!/bin/bash
# AWS user-data script
# Derive a new hostname and set it as the node boots
# -----
NEW_HOSTNAME=$(curl -s http://169.254.169.254/latest/meta-data/public-hostname | awk -F.  '{print $1"." ".{{ dns_zone }}"}')
grep -q "^HOSTNAME=" /etc/sysconfig/network && sed -i -e "s/^HOSTNAME=.*$/HOSTNAME=${NEW_HOSTNAME}/" /etc/sysconfig/network || sed -i -e "$ a\HOSTNAME=${NEW_HOSTNAME}" /etc/sysconfig/network
hostname $NEW_HOSTNAME
hostnamectl_bin="$(which hostnamectl)"
systemctl_bin="$(which systemctl)"

if [ -x "$hostnamectl_bin" ]; then
  $hostnamectl_bin set-hostname ${NEW_HOSTNAME}
  $hostnamectl_bin set-hostname --static ${NEW_HOSTNAME}
  $hostnamectl_bin set-hostname "Example pretty hostname ${NEW_HOSTNAME%%.{{ dns_zone }}}" --pretty
  $hostnamectl_bin set-chassis vm
  $systemctl_bin restart systemd-hostnamed
fi
sed -i -e "s/TEMPLATE/${NEW_HOSTNAME}/" /etc/motd

# Sledgehammer for CentOS 7
cat << EOF > /etc/dhcp/dhclient-eth0.conf
interface "eth0"
  {
    supersede host-name "${NEW_HOSTNAME}";
  }
EOF

echo "${NEW_HOSTNAME}" > /etc/hostname

```

This user-data script is a template using `dns_name` as the Route53 zone name suffix to use.  The original intent for this script was to rename a host using the `ec2-<ip_here>` prefix, and some arbitrary Route53 zone suffix (which could be a custom private zone).

The Ansible tasks attached to this also do some magic to workaround `cloud-init` or `systemd-hostnamed` overwriting the hostname again.  If intended hostname is not detected in `hostname` command output, it will re-run the user data script from `http://169.254.169.254/latest/user-data`.

Here is the example playbook:

- `playbooks/example/launch-ec2-example.yml`:
```
# This playbook launches new Example hosts
- name: Launch new Example host(s)
  hosts: localhost
  connection: local
  gather_facts: false
  vars_files:
    - vars/shared.yml
  vars:
    location: "{{ aws_region|default('us-east-1') }}"
    region: "{{ aws_region|default('us-east-1') }}"
    account: "{{ aws_account|default('efd') }}"
  pre_tasks:
  # Load a variable file based on the AWS account & Region, or just shared default if not found.
    - include_vars: "{{ item }}"
      with_first_found:
        - "{{ alt_config | default('vars/example-us-east-1.yml') }}"
        - "vars/{{ account }}-{{ infra_environment | lower }}-{{ region }}.yml"
        - "vars/{{ account }}-{{ region }}.yml"
        - "vars/{{ region }}.yml"
        - "vars/shared.yml"
      tags:
        - launch_instance
        - wait_for_ssh
        - sshd_race_condition_wait
  tasks:
    - name: Get VPC Net Facts
      ec2_vpc_net_facts:
        filters:
          "tag:Name": "{{ vpc_tags['Name'] }}"
      register: vpc_result

    - name: Gather Subnet Facts for Private
      local_action:
        module: ec2_vpc_subnet_facts
        filters:
          "tag:private": "True"
          vpc-id: "{{ vpc_result['vpcs'][0].id }}"
      when: multi_az is defined and multi_az
      register: vpc_subnets_private
      tags: launch_instance

    - name: Set vpc_subnets variable to Public Subnet result
      set_fact: vpc_subnets="{{ vpc_subnets_private.subnets | sort(attribute='availability_zone') }}"
      check_mode: no
      always_run: yes
      failed_when: false
      ignore_errors: true
      when: multi_az is defined and multi_az and (vpc_subnets_private is defined and vpc_subnets_private.subnets is defined)
      tags: launch_instance

    - name: Launch ExampleHost instance(s)
      local_action:
        module: ec2
        state: present
        count: "{{ count | default(1) }}"
        instance_type: "{{ instance_type }}"
        image: "{% if build_from_base is defined %}{{ instance_base_ami_id }}{% else %}{{ instance_ami_id }}{% endif %}"
        key_name: "{{ key_name }}"
        assign_public_ip: yes
        wait: yes
        wait_timeout: 600
        region: "{{ region }}"
        instance_profile_name: "{{ instance_iam_role }}"
        vpc_subnet_id: "{{ vpc_subnet_id }}"
        group: "{{ security_group_name }}"
        user_data: "{{ lookup('template', playbook_dir + '/templates/user-data.sh.j2') }}"
        instance_tags:
          Name: "examplehost-init"
          app: "examplehost-initializing"
          infra_environment: "{{ infra_environment|default('prod') }}"
          location: '{{ common_tags.location }}'
          component: '{{ common_tags.component }}'
          business_unit: '{{ common_tags.business_unit }}'
          product: '{{ common_tags.product }}'
          support_level: '{{ common_tags.support_level }}'
          os: '{{ common_tags.os }}'
          os_version: '{{ common_tags.os_version }}'
        volumes:
          - device_name: /dev/xvdf
            volume_size: 48
            volume_type: gp2
            delete_on_termination: true
      register: ec2
      tags: launch_instance

    ## Because cloud-init can take a while to run, and is what creates the image_user, 
    ## we must wait for it to exist before we can SSH in successfully. (note: wait_for just checks for if port is listening)
    ## Bug reproduction playbook: https://gist.github.com/trinitronx/afd894c89384d413597b
    - name: Attempt to avoid cloud-init sshd race condition... ssh connect & retry for 5 minutes
      # Default to rpcm unless we include vars/ansible_user.yml later when not building from a clean CentOS base AMI
      remote_user: '{{ ansible_client_user }}'
      local_action: shell ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o BatchMode=yes -o ConnectTimeout={{ ansible_ssh_connect_timeout | default('30')}} -i {{ ansible_ssh_private_key_file }} -p {{ ansible_ssh_port | default('22') }} {{ ansible_user | default(ansible_client_user) }}@{{ item.private_ip }} echo success
      register: ssh_as_ansible_user
      failed_when:  ssh_as_ansible_user.stdout.find('failed') != -1
      changed_when: ssh_as_ansible_user.stderr.find('Permission denied') == -1
      until: ssh_as_ansible_user.stdout.find("success") != -1
      retries: 600
      delay: 1
      with_items: "{{ ec2.instances | list }}"
      tags:
        - wait_for_ssh
        - sshd_race_condition_wait

    - name: Re-tag ExampleHost instances with hostname
      ec2_tag: >
        resource="{{ item.id }}"
        region="{{ region }}"
        state=present
      with_items: "{{ ec2.instances | list }}"
      args:
        tags:
          Name: "ec2-{{ item.public_ip|replace('.','-') }}.{{ dns_zone }}"
          location: "{{ region }}"
          roles: "example"
      delegate_to: 127.0.0.1
      tags:
        - launch_instance
        - tag_launched_hosts
        - ec2_tag

    - name: Add DNS A record for new instance
      local_action:
        module: route53
        command: create
        overwrite: yes
        zone: "{{ dns_zone }}"
        hosted_zone_id: "{{ route53_primary_zone_id }}"
        private_zone: "{{ route53_primary_zone_is_private }}"
        record: "ec2-{{ item.public_ip|replace('.','-') }}.{{ dns_zone }}"
        type: A
        ttl: 300
        value: "{{ item.public_ip }}"
      with_items: "{{ ec2.instances | list }}"
      tags:
        - route53

# Workaround for cloud-init bug: https://pad.lv/bug=1246485
# Avoid having cloud-init reset the hostname which user-data.sh sets
    - name: Workaround for cloud-init #1246485
      lineinfile: >
        dest="/etc/cloud/cloud.cfg"
        line="{{ item.line }}"
        regexp="{{ item.pattern }}"
        validate="python -c 'import yaml; print yaml.load(  open(\"%s\", \"r\").read() )'"
      with_items:
        - { line: '# - set_hostname', pattern: '^\s*-\s*set_hostname' }
        - { line: '# - update_hostname', pattern: '^\s*-\s*update_hostname' }

    - name: Check that hostname includes "ec2-public-ip"
      shell: >
        hostname | grep -q ec2-{{ ec2_ip_address | replace('.','-') }}.{{ dns_zone }}
      register: example_hostname_result
      check_mode: no
      always_run: yes
      failed_when: false

    - name: Re-run user-data.sh script to set hostname ONCE AND FOR ALL!
      shell: >
        curl -L http://169.254.169.254/latest/user-data | bash -x
      when: example_hostname_result.rc != 0

```

- Vars file for `aws_account` + `infra_environment` + `aws_region` triplet =  `playbooks/example/vars/example-dev-us-east-1.yml`:

**Fill this with your data!**

```
---
# Variables for YOUR Amazon account: us-east-1 Example hosts
instance_type: 'c4.large'
region: "us-east-1"
zone: "us-east-1e"

instance_base_ami_id: ami-46c1b650 ## CentOS 7 (x86_64) with Updates HVM (1704_01 2017-05-15, released 05/15/2017)
instance_ami_id:  ami-xxxxxxxx ## Your pre-built ExampleHost base AMI here
image_os: "CentOS"
image_os_version: 7

instance_profile_name: "example"
instance_iam_role: 'example'
security_group_name: 'example'

image_user: centos                 ## Default user for AMI

# Subnets for single-AZ deployment
### PUT YOUR SUBNET IDs HERE
# EFD DEV Public E
subnet_id_public: &VPCSUBNETPUB "subnet-xxxxxxxx"
# EFD DEV Private E
subnet_id_private: &VPCSUBNETPRIV "subnet-xxxxxxxx"

vpc_subnet_id: *VPCSUBNETPUB

## Override your default ansible_client_user here
ansible_client_user: centos
# ConnectTimeout setting for SSH connection tests (cloud-init race condition workaround)
# Set this higher if you are connecting from a slower connection (e.g.: VPN)
ansible_ssh_connect_timeout: 20
# Private key file here
private_key_file: '~/.ssh/example.pem'
ansible_ssh_private_key_file: "{{ (private_key_file | bool) | ternary(private_key_file, omit) }}"

instance_root_device: '/dev/sda1'
key_name: 'example'
keypair: 'example'

## PUT YOUR PUBLIC / PRIVATE SecurityGroup IDs here
# Example-Default-SG
security_group_public: "sg-xxxxxxxx"
# Example-Private-NAT-SG
security_group_private: "sg-xxxxxxxx"

## PUT YOUR DNS ZONE ID + NAME HERE
# Lock any Route53 changes to the hosted zone ID for EFD-DEV-VPC
route53_primary_zone_id: ZXXXXXXXXXXXXX
route53_primary_zone_is_private: true
# DNS Zone the EMX hosts live in
dns_zone: local.example


### PUT YOUR DESIRED EC2 INSTANCE TAGS HERE
common_tags: &COMMON_TAGS
  location: '{{ region }}'
  component: 'infrastructure'
  business_unit: 'AcmeCo'
  product: 'Anvil'
  infra_environment: '{{ infra_environment | lower }}'
  support_level: '{{ infra_environment | lower }}'
  os: '{{ image_os }}'
  os_version: '{{ image_os_version }}'

### PUT YOUR VPC TAGS HERE
## The playbook task will detect VPC ID from the Name tag
# VPC information
vpc_cidr_block: &VPCCIDR 10.100.1.0/24
vpc_tags: &VPC_TAGS
  Name: EXAMPLE-VPC
  support_level: dev
  component: infrastructure
  business_unit: AcmeCo
  product: Anvil
  location: us-east-1

```

**Comment by ptux** [comment]
This problem bothers me too.

Here is my idea.

- we add a bool option `reboot`
- if `reboot=yes`, we check the hostname after reboot.
- if NOT match, we tell user with failed.

**Comment by sivel** [comment]
Closing per https://github.com/ansible/ansible/issues/19711#issuecomment-271157611
